### DISCUSSION TITLE
I have problems with selecting data from store

### DISCUSSION THREAD

**Discussion by samik3k** [pmndrs/zustand#1917] [Q&A] [answered]
So i read https://docs.pmnd.rs/zustand/recipes/recipes#selecting-multiple-state-slices

package.json
```
"next": "^13.4.7",
"ramda": "^0.28.0",
"react": "18.2.0",
"zustand": "^4.3.6"
```

nodejs 16.18.1

I have store something like
``` typescript
type TList = { [key: string]: { id: string; name: string; } }
type TItem = { [key: string]: { id: string; data: string[]; someField: any; listId: string; } }

type TStore = {
  list: TList;
  item: TItem;
  something: any;
}

const store = create<TStore>()({
  list: {},
  item: {},
  something: null,
})
```

I made custom hook, like
``` typescript
const useGetSomething = () => {
  const {
    list,
    item,
  } = useStore(state => ({
    list: state.list,
    item: state.item,
  }))
  
  // other operations with list and item
  
  return someObject
}
```
With that example i will get rerender in my component that using useGetSomething() when someone updates any attributes in store.
So i added **shallow**, but it didnt work for some reasons.
``` typescript
const {
  list,
  item,
} = useStore(state => ({
  list: state.list,
  item: state.item,
}), shallow)
```
I still dont get any rerender in my component.

I was trying to use:
``` typescript
const list = useStore(state => state.list)
const item = useStore(state => state.item)
```
And this:
``` typescript
const item = useStore(state => state.item, (a, b) => equals(a, b))
```
I getting always true when comparing a and b
I even was tried `JSON.stringify(a) === JSON.stringify(b)` for comparing 

And this:
``` typescript
const item = useStore(state => state.item, () => false)
```
Still no result => rerender when data is changed.
Wanted result get component rerender on list and item updates only.

I'm doing something wrong?

**Comment by samik3k** [comment]
Its seems that im getting same value for a and b in comparison...

**Comment by dai-shi** [comment]
I'd suggest to try creating a very small example with https://codesandbox.io/s/react-new.

**Reply by samik3k** [reply]
Yes, i will try today to make example

**Reply by samik3k** [reply]
I made example https://codesandbox.io/s/cocky-panka-kj6hdh?file=/src/App.tsx

**Reply by samik3k** [reply]
I made one more example https://codesandbox.io/s/frosty-noether-d7m446?file=/src/App.tsx
So if it's a simple boolean, its working fine

**Reply by samik3k** [reply]
With string or number it's working too...

**Reply by samik3k** [reply]
And it ignores `return false` in equality function for arrays and objects

**Reply by dai-shi** [reply, marked as answer]
https://codesandbox.io/s/xenodochial-hellman-dnks99
You need to update state immutably for nested objects too.

**Reply by samik3k** [reply]
Hmmmmm

**Reply by samik3k** [reply]
Thx, its working now!
I was looking what i was doing wrong
Thanks for help!

P.S. nice store library :)
